### PR TITLE
move CUDAStream to phi

### PR DESCRIPTION
**Pull request by zhiqiu**
### PR types

Bug fixes
### PR changes

Others
### Describe

move CUDAStream to phi

The original implementation of CUDADeviceContext contains 2 instances of  gpuStream, one in cudaStream and one inherits from GPUContext, which results in data inconsistent, i.e, developer may write one stream and read another.

This PR fixes the bug and moves `CUDAStream` to `GPUContext`.

ps, this pr is the preparatory work of unifing CUDADeviceContext and GPUContext.

### CODE REVIEW

File: paddle/phi/backends/gpu/gpu_context.cc [hunk 1]
```diff
@@ -740,20 +777,26 @@ struct GPUContext::Impl {
...
+}
 
 GPUContext::GPUContext(GPUContext&&) = default;
 
 GPUContext& GPUContext::operator=(GPUContext&&) = default;
 
 GPUContext::GPUContext(const GPUPlace& place)
-    : DeviceContext(), impl_(std::make_unique<Impl>(place)) {}
+    : DeviceContext(), impl_(std::make_unique<Impl>(place)) {
+  impl_->PartialInitWithoutAllocator();
```
Comment (jiweibo) [inline paddle/phi/backends/gpu/gpu_context.cc]: 推理的场景需要构造一个空的GPUContext，后续由用户决定哪些资源初始化，构造中的Init能否去掉呢？

Comment (zhiqiu) [reply paddle/phi/backends/gpu/gpu_context.cc]: 不能去掉，训练场景需要构造，要不加一个bool参数？

File: paddle/phi/backends/gpu/gpu_context.cc [hunk 2]
```diff
@@ -212,12 +213,14 @@ struct GPUContext::Impl {
                            &max_threads_per_mp_,
                            &max_threads_per_block_,
                            &max_grid_dim_size_);
-    phi::InitStream(&stream_);
+    stream_ = new CUDAStream(place_);
+    InitEigenDevice();
```
Comment (jiweibo) [inline paddle/phi/backends/gpu/gpu_context.cc:217]: eigen device好像已经切换成lazy初始化了，后面有空的话可以删掉

**Review by jiweibo — APPROVED** [review]
LGTM

**Comment by jiweibo** [comment]
> LGTM

TODO: @jiweibo inference context should construct without any resource.

**Review by zyfncg — APPROVED** [review]

**Review by qili93 — APPROVED** [review]
LGTM for tensor.h

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**